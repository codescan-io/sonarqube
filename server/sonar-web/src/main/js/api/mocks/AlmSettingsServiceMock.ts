/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { mockAlmSettingsInstance } from '../../helpers/mocks/alm-settings';
import { AlmKeys, AlmSettingsInstance } from '../../types/alm-settings';
import { getAlmSettings } from '../alm-settings';

export default class AlmSettingsServiceMock {
  almSettings: AlmSettingsInstance[];
  defaultSetting: AlmSettingsInstance[] = [
    mockAlmSettingsInstance({ key: 'conf-final-1', alm: AlmKeys.GitLab }),
    mockAlmSettingsInstance({ key: 'conf-final-2', alm: AlmKeys.GitLab }),
    mockAlmSettingsInstance({ key: 'conf-github-1', alm: AlmKeys.GitHub, url: 'http://url' }),
    mockAlmSettingsInstance({ key: 'conf-github-2', alm: AlmKeys.GitHub, url: 'http://url' }),
    mockAlmSettingsInstance({ key: 'conf-github-3', alm: AlmKeys.GitHub, url: 'javascript://url' }),
    mockAlmSettingsInstance({ key: 'conf-azure-1', alm: AlmKeys.Azure, url: 'url' }),
    mockAlmSettingsInstance({ key: 'conf-azure-2', alm: AlmKeys.Azure, url: 'url' }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketcloud-1',
      alm: AlmKeys.BitbucketCloud,
      url: 'url',
    }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketcloud-2',
      alm: AlmKeys.BitbucketCloud,
      url: 'url',
    }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketserver-1',
      alm: AlmKeys.BitbucketServer,
      url: 'url',
    }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketserver-2',
      alm: AlmKeys.BitbucketServer,
      url: 'url',
    }),
  ];

  constructor() {
    this.almSettings = cloneDeep(this.defaultSetting);
    jest.mocked(getAlmSettings).mockImplementation(this.getAlmSettingsHandler);
  }

  getAlmSettingsHandler = () => {
    return Promise.resolve(this.almSettings);
  };

  reset = () => {
    this.almSettings = cloneDeep(this.defaultSetting);
  };
}
