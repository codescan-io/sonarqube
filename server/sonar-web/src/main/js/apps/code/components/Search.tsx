/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isEmpty, omit } from 'lodash';
import * as React from 'react';
import { getTree } from '../../../api/components';
import ButtonToggle from '../../../components/controls/ButtonToggle';
import SearchBox from '../../../components/controls/SearchBox';
import { Location, Router, withRouter } from '../../../components/hoc/withRouter';
import DeferredSpinner from '../../../components/ui/DeferredSpinner';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { KeyboardKeys } from '../../../helpers/keycodes';
import { translate } from '../../../helpers/l10n';
import { BranchLike } from '../../../types/branch-like';
import { ComponentQualifier } from '../../../types/component';
import { ComponentMeasure } from '../../../types/types';

interface Props {
  branchLike?: BranchLike;
  component: ComponentMeasure;
  location: Location;
  newCodeSelected: boolean;
  onSearchClear: () => void;
  onNewCodeToggle: (newCode: boolean) => void;
  onSearchResults: (results?: ComponentMeasure[]) => void;
  router: Router;
}

interface State {
  query: string;
  loading: boolean;
}

export class Search extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = {
    query: '',
    loading: false,
  };

  componentDidMount() {
    this.mounted = true;
    if (this.props.location.query.search) {
      this.handleQueryChange(this.props.location.query.search);
    }
  }

  componentDidUpdate(nextProps: Props) {
    // if the component has change, reset the current state
    if (nextProps.location.query.id !== this.props.location.query.id) {
      this.setState({
        query: '',
        loading: false,
      });
      this.props.onSearchClear();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    switch (event.nativeEvent.key) {
      case KeyboardKeys.Enter:
      case KeyboardKeys.UpArrow:
      case KeyboardKeys.DownArrow:
        event.preventDefault();
        event.currentTarget.blur();
        break;
    }
  };

  handleSearch = (query: string) => {
    if (this.mounted) {
      const { branchLike, component, router, location } = this.props;
      this.setState({ loading: true });

      if (query !== location.query.search) {
        router.replace({
          pathname: location.pathname,
          query: { ...location.query, search: query },
        });
      }

      const qualifiers = [
        ComponentQualifier.Portfolio,
        ComponentQualifier.SubPortfolio,
        ComponentQualifier.Application,
      ].includes(component.qualifier as ComponentQualifier)
        ? [ComponentQualifier.SubPortfolio, ComponentQualifier.Project].join(',')
        : [ComponentQualifier.TestFile, ComponentQualifier.File].join(',');

      getTree({
        component: component.key,
        q: query,
        s: 'qualifier,name',
        qualifiers,
        ...getBranchLikeQuery(branchLike),
      })
        .then((r) => {
          if (this.mounted) {
            this.setState({
              loading: false,
            });
            this.props.onSearchResults(r.components);
          }
        })
        .catch(() => {
          if (this.mounted) {
            this.setState({ loading: false });
          }
        });
    }
  };

  handleQueryChange = (query: string) => {
    const { router, location } = this.props;
    this.setState({ query });
    if (query.length === 0) {
      router.replace({ pathname: location.pathname, query: omit(location.query, 'search') });
      this.props.onSearchClear();
    } else {
      this.handleSearch(query);
    }
  };

  render() {
    const { component, newCodeSelected } = this.props;
    const { loading, query } = this.state;
    const isPortfolio = ['VW', 'SVW', 'APP'].includes(component.qualifier);

    return (
      <div className="code-search" id="code-search">
        {isPortfolio && (
          <span className="big-spacer-right">
            <ButtonToggle
              disabled={!isEmpty(query)}
              options={[
                {
                  value: true,
                  label: translate('projects.view.new_code'),
                },
                {
                  value: false,
                  label: translate('projects.view.overall_code'),
                },
              ]}
              value={newCodeSelected}
              onCheck={this.props.onNewCodeToggle}
            />
          </span>
        )}
        <SearchBox
          minLength={3}
          onChange={this.handleQueryChange}
          onKeyDown={this.handleKeyDown}
          placeholder={translate(
            isPortfolio ? 'code.search_placeholder.portfolio' : 'code.search_placeholder'
          )}
          value={this.state.query}
        />
        <DeferredSpinner className="spacer-left" loading={loading} />
      </div>
    );
  }
}

export default withRouter(Search);
