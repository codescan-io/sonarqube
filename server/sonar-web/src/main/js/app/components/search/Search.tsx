/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { debounce, uniqBy } from 'lodash';
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { getSuggestions } from '../../../api/components';
import { DropdownOverlay } from '../../../components/controls/Dropdown';
import FocusOutHandler from '../../../components/controls/FocusOutHandler';
import OutsideClickHandler from '../../../components/controls/OutsideClickHandler';
import SearchBox from '../../../components/controls/SearchBox';
import { Router, withRouter } from '../../../components/hoc/withRouter';
import ClockIcon from '../../../components/icons/ClockIcon';
import DeferredSpinner from '../../../components/ui/DeferredSpinner';
import { isInput, isShortcut } from '../../../helpers/keyboardEventHelpers';
import { KeyboardKeys } from '../../../helpers/keycodes';
import { translate, translateWithParameters } from '../../../helpers/l10n';
import { getKeyboardShortcutEnabled } from '../../../helpers/preferences';
import { scrollToElement } from '../../../helpers/scrolling';
import { getComponentOverviewUrl } from '../../../helpers/urls';
import { ComponentQualifier } from '../../../types/component';
import { Dict } from '../../../types/types';
import RecentHistory from '../RecentHistory';
import './Search.css';
import SearchResult from './SearchResult';
import SearchResults from './SearchResults';
import { ComponentResult, More, Results, sortQualifiers } from './utils';

interface Props {
  router: Router;
}
interface State {
  loading: boolean;
  loadingMore?: string;
  more: More;
  open: boolean;
  query: string;
  results: Results;
  selected?: string;
  shortQuery: boolean;
}

const MIN_SEARCH_QUERY_LENGTH = 2;

export class Search extends React.PureComponent<Props, State> {
  input?: HTMLInputElement | null;
  node?: HTMLElement | null;
  nodes: Dict<HTMLElement>;
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.nodes = {};
    this.search = debounce(this.search, 250);
    this.state = {
      loading: false,
      more: {},
      open: false,
      query: '',
      results: {},
      shortQuery: false,
    };
  }

  componentDidMount() {
    this.mounted = true;
    document.addEventListener('keydown', this.handleKeyDown);
    document.addEventListener('keydown', this.handleSKeyDown);
  }

  componentDidUpdate(_prevProps: Props, prevState: State) {
    if (prevState.selected !== this.state.selected) {
      this.scrollToSelected();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
    document.removeEventListener('keydown', this.handleSKeyDown);
    document.removeEventListener('keydown', this.handleKeyDown);
  }

  focusInput = () => {
    if (this.input) {
      this.input.focus();
    }
  };

  handleClickOutside = () => {
    this.closeSearch(false);
  };

  handleFocus = () => {
    if (!this.state.open) {
      // simulate click to close any other dropdowns
      const body = document.documentElement;
      if (body) {
        body.click();
      }
    }
    this.openSearch();
  };

  openSearch = () => {
    if (!this.state.open && !this.state.query) {
      this.search('');
    }
    this.setState({ open: true });
  };

  closeSearch = (clear = true) => {
    if (this.input) {
      this.input.blur();
    }
    if (clear) {
      this.setState({
        more: {},
        open: false,
        query: '',
        results: {},
        selected: undefined,
        shortQuery: false,
      });
    } else {
      this.setState({ open: false });
    }
  };

  getPlainComponentsList = (results: Results, more: More) =>
    sortQualifiers(Object.keys(results)).reduce((components, qualifier) => {
      const next = [...components, ...results[qualifier].map((component) => component.key)];
      if (more[qualifier]) {
        next.push('qualifier###' + qualifier);
      }
      return next;
    }, []);

  stopLoading = () => {
    if (this.mounted) {
      this.setState({ loading: false });
    }
  };

  search = (query: string) => {
    if (query.length === 0 || query.length >= MIN_SEARCH_QUERY_LENGTH) {
      this.setState({ loading: true });
      const recentlyBrowsed = RecentHistory.get().map((component) => component.key);
      getSuggestions(query, recentlyBrowsed).then((response) => {
        // compare `this.state.query` and `query` to handle two request done almost at the same time
        // in this case only the request that matches the current query should be taken
        if (this.mounted && this.state.query === query) {
          const results: Results = {};
          const more: More = {};
          this.nodes = {};
          response.results.forEach((group) => {
            results[group.q] = group.items.map((item) => ({ ...item, qualifier: group.q }));
            more[group.q] = group.more;
          });
          const list = this.getPlainComponentsList(results, more);
          this.setState({
            loading: false,
            more,
            results,
            selected: list.length > 0 ? list[0] : undefined,
            shortQuery:
              query.length > MIN_SEARCH_QUERY_LENGTH && response.warning === 'short_input',
          });
        }
      }, this.stopLoading);
    } else {
      this.setState({ loading: false });
    }
  };

  searchMore = (qualifier: string) => {
    const { query } = this.state;
    if (query.length === 1) {
      return;
    }

    this.setState({ loading: true, loadingMore: qualifier });
    const recentlyBrowsed = RecentHistory.get().map((component) => component.key);
    getSuggestions(query, recentlyBrowsed, qualifier).then((response) => {
      if (this.mounted) {
        const group = response.results.find((group) => group.q === qualifier);
        const moreResults = (group ? group.items : []).map((item) => ({ ...item, qualifier }));
        this.setState((state) => ({
          loading: false,
          loadingMore: undefined,
          more: { ...state.more, [qualifier]: 0 },
          results: {
            ...state.results,
            [qualifier]: uniqBy([...state.results[qualifier], ...moreResults], 'key'),
          },
          selected: moreResults.length > 0 ? moreResults[0].key : state.selected,
        }));
        this.focusInput();
      }
    }, this.stopLoading);
  };

  handleQueryChange = (query: string) => {
    this.setState({ query, shortQuery: query.length === 1 });
    this.search(query);
  };

  selectPrevious = () => {
    this.setState(({ more, results, selected }) => {
      if (selected) {
        const list = this.getPlainComponentsList(results, more);
        const index = list.indexOf(selected);
        return index > 0 ? { selected: list[index - 1] } : null;
      }
      return null;
    });
  };

  selectNext = () => {
    this.setState(({ more, results, selected }) => {
      if (selected) {
        const list = this.getPlainComponentsList(results, more);
        const index = list.indexOf(selected);
        return index >= 0 && index < list.length - 1 ? { selected: list[index + 1] } : null;
      }
      return null;
    });
  };

  openSelected = () => {
    const { results, selected } = this.state;

    if (!selected) {
      return;
    }

    if (selected.startsWith('qualifier###')) {
      this.searchMore(selected.substr(12));
    } else {
      let qualifier = ComponentQualifier.Project;

      if ((results[ComponentQualifier.Portfolio] ?? []).find((r) => r.key === selected)) {
        qualifier = ComponentQualifier.Portfolio;
      } else if ((results[ComponentQualifier.SubPortfolio] ?? []).find((r) => r.key === selected)) {
        qualifier = ComponentQualifier.SubPortfolio;
      }

      this.props.router.push(getComponentOverviewUrl(selected, qualifier));

      this.closeSearch();
    }
  };

  scrollToSelected = () => {
    if (this.state.selected) {
      const node = this.nodes[this.state.selected];
      if (node && this.node) {
        scrollToElement(node, { topOffset: 30, bottomOffset: 30, parent: this.node });
      }
    }
  };

  handleSKeyDown = (event: KeyboardEvent) => {
    if (!getKeyboardShortcutEnabled() || isInput(event) || isShortcut(event)) {
      return true;
    }
    if (event.key === KeyboardKeys.KeyS) {
      event.preventDefault();
      this.focusInput();
      this.openSearch();
    }
  };

  handleKeyDown = (event: KeyboardEvent) => {
    if (!this.state.open) {
      return;
    }

    switch (event.key) {
      case KeyboardKeys.Enter:
        event.preventDefault();
        event.stopPropagation();
        this.openSelected();
        break;
      case KeyboardKeys.UpArrow:
        event.preventDefault();
        event.stopPropagation();
        this.selectPrevious();
        break;
      case KeyboardKeys.Escape:
        event.preventDefault();
        event.stopPropagation();
        this.closeSearch();
        break;
      case KeyboardKeys.DownArrow:
        event.preventDefault();
        event.stopPropagation();
        this.selectNext();
        break;
    }
  };

  handleSelect = (selected: string) => {
    this.setState({ selected });
  };

  innerRef = (component: string, node: HTMLElement | null) => {
    if (node) {
      this.nodes[component] = node;
    }
  };

  searchInputRef = (node: HTMLInputElement | null) => {
    this.input = node;
  };

  renderResult = (component: ComponentResult) => (
    <SearchResult
      component={component}
      innerRef={this.innerRef}
      key={component.key}
      onClose={this.closeSearch}
      onSelect={this.handleSelect}
      selected={this.state.selected === component.key}
    />
  );

  renderNoResults = () => (
    <div className="navbar-search-no-results" aria-live="assertive">
      {translateWithParameters('no_results_for_x', this.state.query)}
    </div>
  );

  render() {
    const search = (
      <div className="navbar-search dropdown">
        <DeferredSpinner className="navbar-search-icon" loading={this.state.loading} />

        <SearchBox
          autoFocus={this.state.open}
          innerRef={this.searchInputRef}
          minLength={2}
          onChange={this.handleQueryChange}
          onFocus={this.handleFocus}
          placeholder={translate('search.placeholder')}
          value={this.state.query}
        />

        {this.state.shortQuery && (
          <span className="navbar-search-input-hint" aria-live="assertive">
            {translateWithParameters('select2.tooShort', MIN_SEARCH_QUERY_LENGTH)}
          </span>
        )}

        {this.state.open && Object.keys(this.state.results).length > 0 && (
          <DropdownOverlay noPadding={true}>
            <div className="global-navbar-search-dropdown" ref={(node) => (this.node = node)}>
              <SearchResults
                allowMore={this.state.query.length !== 1}
                loadingMore={this.state.loadingMore}
                more={this.state.more}
                onMoreClick={this.searchMore}
                onSelect={this.handleSelect}
                renderNoResults={this.renderNoResults}
                renderResult={this.renderResult}
                results={this.state.results}
                selected={this.state.selected}
              />
              <div className="dropdown-bottom-hint">
                <div className="pull-right" aria-hidden={true}>
                  <ClockIcon className="little-spacer-right" size={12} />
                  {translate('recently_browsed')}
                </div>
                <FormattedMessage
                  defaultMessage={translate('search.shortcut_hint')}
                  id="search.shortcut_hint"
                  values={{
                    shortcut: <span className="shortcut-button shortcut-button-small">s</span>,
                  }}
                />
              </div>
            </div>
          </DropdownOverlay>
        )}
      </div>
    );

    return this.state.open ? (
      <FocusOutHandler onFocusOut={this.handleClickOutside}>
        <OutsideClickHandler onClickOutside={this.handleClickOutside}>{search}</OutsideClickHandler>
      </FocusOutHandler>
    ) : (
      search
    );
  }
}

export default withRouter(Search);
