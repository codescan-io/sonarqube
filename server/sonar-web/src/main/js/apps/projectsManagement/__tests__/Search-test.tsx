/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { mockReactSelectOptionProps } from '../../../helpers/mocks/react-select';
import { mockAppState } from '../../../helpers/testMocks';
import { click } from '../../../helpers/testUtils';
import { Props, Search } from '../Search';

it('renders', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('disables the delete and bulk apply buttons unless a project is selected', () => {
  const wrapper = shallowRender();
  expect(wrapper.find('Button.js-delete').prop('disabled')).toBe(true);
  expect(wrapper.find('Button.js-bulk-apply-permission-template').prop('disabled')).toBe(true);

  wrapper.setProps({ selection: ['foo'] });
  expect(wrapper.find('Button.js-delete').prop('disabled')).toBe(false);
  expect(wrapper.find('Button.js-bulk-apply-permission-template').prop('disabled')).toBe(false);
});

it('render qualifiers filter', () => {
  expect(
    shallowRender({ appState: mockAppState({ qualifiers: ['TRK', 'VW', 'APP'] }) })
  ).toMatchSnapshot();
});

it('updates qualifier', () => {
  const onQualifierChanged = jest.fn();
  const wrapper = shallowRender({
    onQualifierChanged,
    appState: mockAppState({ qualifiers: ['TRK', 'VW', 'APP'] }),
  });
  wrapper.find('Select[name="projects-qualifier"]').simulate('change', {
    value: 'VW',
  });
  expect(onQualifierChanged).toHaveBeenCalledWith('VW');
});

it('renders optionrenderer and singlevaluerenderer', () => {
  const wrapper = shallowRender({
    appState: mockAppState({ qualifiers: ['TRK', 'VW', 'APP'] }),
  });
  const OptionRendererer = wrapper.instance().optionRenderer;
  const SingleValueRendererer = wrapper.instance().singleValueRenderer;
  expect(
    shallow(<OptionRendererer {...mockReactSelectOptionProps({ value: 'val' })} />)
  ).toMatchSnapshot('option renderer');
  expect(
    shallow(<SingleValueRendererer {...mockReactSelectOptionProps({ value: 'val' })} />)
  ).toMatchSnapshot('single value renderer');
});

it('selects provisioned', () => {
  const onProvisionedChanged = jest.fn();
  const wrapper = shallowRender({ onProvisionedChanged });
  wrapper.find('Checkbox[id="projects-provisioned"]').prop<Function>('onCheck')(true);
  expect(onProvisionedChanged).toHaveBeenCalledWith(true);
});

it('does not render provisioned filter for portfolios', () => {
  const wrapper = shallowRender();
  expect(wrapper.find('Checkbox[id="projects-provisioned"]').exists()).toBe(true);
  wrapper.setProps({ qualifiers: 'VW' });
  expect(wrapper.find('Checkbox[id="projects-provisioned"]').exists()).toBe(false);
});

it('updates analysis date', () => {
  const onDateChanged = jest.fn();
  const wrapper = shallowRender({ onDateChanged });

  wrapper.find('DateInput').prop<Function>('onChange')('2017-04-08T00:00:00.000Z');
  expect(onDateChanged).toHaveBeenCalledWith('2017-04-08T00:00:00.000Z');

  wrapper.find('DateInput').prop<Function>('onChange')(undefined);
  expect(onDateChanged).toHaveBeenCalledWith(undefined);
});

it('searches', () => {
  const onSearch = jest.fn();
  const wrapper = shallowRender({ onSearch });
  wrapper.find('SearchBox').prop<Function>('onChange')('foo');
  expect(onSearch).toHaveBeenCalledWith('foo');
});

it('checks all or none projects', () => {
  const onAllDeselected = jest.fn();
  const onAllSelected = jest.fn();
  const wrapper = shallowRender({ onAllDeselected, onAllSelected });

  wrapper.find('Checkbox[id="projects-selection"]').prop<Function>('onCheck')(true);
  expect(onAllSelected).toHaveBeenCalled();

  wrapper.find('Checkbox[id="projects-selection"]').prop<Function>('onCheck')(false);
  expect(onAllDeselected).toHaveBeenCalled();
});

it('deletes projects', () => {
  const onDeleteProjects = jest.fn();
  const wrapper = shallowRender({ onDeleteProjects, selection: ['foo', 'bar'] });
  click(wrapper.find('.js-delete'));
  expect(wrapper.find('DeleteModal')).toMatchSnapshot();
  wrapper.find('DeleteModal').prop<Function>('onConfirm')();
  expect(onDeleteProjects).toHaveBeenCalled();
});

it('bulk applies permission template', () => {
  const wrapper = shallowRender({});
  click(wrapper.find('.js-bulk-apply-permission-template'));
  expect(wrapper.find('BulkApplyTemplateModal')).toMatchSnapshot();
  wrapper.find('BulkApplyTemplateModal').prop<Function>('onClose')();
  wrapper.update();
  expect(wrapper.find('BulkApplyTemplateModal').exists()).toBe(false);
});

function shallowRender(props?: { [P in keyof Props]?: Props[P] }) {
  return shallow<Search>(
    <Search
      analyzedBefore={undefined}
      onAllDeselected={jest.fn()}
      onAllSelected={jest.fn()}
      onDateChanged={jest.fn()}
      onDeleteProjects={jest.fn()}
      onProvisionedChanged={jest.fn()}
      onQualifierChanged={jest.fn()}
      onSearch={jest.fn()}
      onVisibilityChanged={jest.fn()}
      projects={[]}
      provisioned={false}
      qualifiers="TRK"
      query=""
      ready={true}
      selection={[]}
      appState={mockAppState({ qualifiers: ['TRK'] })}
      total={17}
      {...props}
    />
  );
}
