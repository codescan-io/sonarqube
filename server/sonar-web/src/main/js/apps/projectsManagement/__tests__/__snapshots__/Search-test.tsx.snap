// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bulk applies permission template 1`] = `
<BulkApplyTemplateModal
  onClose={[Function]}
  provisioned={false}
  qualifier="TRK"
  query=""
  selection={[]}
  total={17}
/>
`;

exports[`deletes projects 1`] = `
<DeleteModal
  onClose={[Function]}
  onConfirm={[Function]}
  provisioned={false}
  qualifier="TRK"
  query=""
  selection={
    [
      "foo",
      "bar",
    ]
  }
  total={17}
/>
`;

exports[`render qualifiers filter 1`] = `
<div
  className="big-spacer-bottom"
>
  <table
    className="data"
  >
    <tbody>
      <tr>
        <td
          className="thin text-middle"
        >
          <Checkbox
            checked={false}
            id="projects-selection"
            onCheck={[Function]}
            thirdState={false}
            title="check_all"
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <Select
            className="input-medium it__project-qualifier-select"
            components={
              {
                "Option": [Function],
                "SingleValue": [Function],
              }
            }
            isDisabled={false}
            isSearchable={false}
            name="projects-qualifier"
            onChange={[Function]}
            options={
              [
                {
                  "label": "qualifiers.TRK",
                  "value": "TRK",
                },
                {
                  "label": "qualifiers.VW",
                  "value": "VW",
                },
                {
                  "label": "qualifiers.APP",
                  "value": "APP",
                },
              ]
            }
            value={
              {
                "label": "qualifiers.TRK",
                "value": "TRK",
              }
            }
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <DateInput
            inputClassName="input-medium"
            name="analyzed-before"
            onChange={[MockFunction]}
            placeholder="last_analysis_before"
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <Select
            className="input-small"
            isDisabled={false}
            isSearchable={false}
            name="projects-visibility"
            onChange={[Function]}
            options={
              [
                {
                  "label": "visibility.both",
                  "value": "all",
                },
                {
                  "label": "visibility.public",
                  "value": "public",
                },
                {
                  "label": "visibility.private",
                  "value": "private",
                },
              ]
            }
            value={
              {
                "label": "visibility.both",
                "value": "all",
              }
            }
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <Checkbox
            checked={false}
            className="link-checkbox-control"
            id="projects-provisioned"
            onCheck={[MockFunction]}
            thirdState={false}
          >
            <span
              className="text-middle little-spacer-left"
            >
              provisioning.only_provisioned
            </span>
          </Checkbox>
          <HelpTooltip
            className="spacer-left"
            overlay="provisioning.only_provisioned.tooltip"
          />
        </td>
        <td
          className="text-middle"
        >
          <SearchBox
            minLength={3}
            onChange={[MockFunction]}
            placeholder="search.search_by_name_or_key"
            value=""
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <Button
            className="js-bulk-apply-permission-template"
            disabled={true}
            onClick={[Function]}
          >
            permission_templates.bulk_apply_permission_template
          </Button>
          <Button
            className="js-delete spacer-left button-red"
            disabled={true}
            onClick={[Function]}
            title="permission_templates.select_to_delete"
          >
            delete
          </Button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders 1`] = `
<div
  className="big-spacer-bottom"
>
  <table
    className="data"
  >
    <tbody>
      <tr>
        <td
          className="thin text-middle"
        >
          <Checkbox
            checked={false}
            id="projects-selection"
            onCheck={[Function]}
            thirdState={false}
            title="check_all"
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <DateInput
            inputClassName="input-medium"
            name="analyzed-before"
            onChange={[MockFunction]}
            placeholder="last_analysis_before"
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <Select
            className="input-small"
            isDisabled={false}
            isSearchable={false}
            name="projects-visibility"
            onChange={[Function]}
            options={
              [
                {
                  "label": "visibility.both",
                  "value": "all",
                },
                {
                  "label": "visibility.public",
                  "value": "public",
                },
                {
                  "label": "visibility.private",
                  "value": "private",
                },
              ]
            }
            value={
              {
                "label": "visibility.both",
                "value": "all",
              }
            }
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <Checkbox
            checked={false}
            className="link-checkbox-control"
            id="projects-provisioned"
            onCheck={[MockFunction]}
            thirdState={false}
          >
            <span
              className="text-middle little-spacer-left"
            >
              provisioning.only_provisioned
            </span>
          </Checkbox>
          <HelpTooltip
            className="spacer-left"
            overlay="provisioning.only_provisioned.tooltip"
          />
        </td>
        <td
          className="text-middle"
        >
          <SearchBox
            minLength={3}
            onChange={[MockFunction]}
            placeholder="search.search_by_name_or_key"
            value=""
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <Button
            className="js-bulk-apply-permission-template"
            disabled={true}
            onClick={[Function]}
          >
            permission_templates.bulk_apply_permission_template
          </Button>
          <Button
            className="js-delete spacer-left button-red"
            disabled={true}
            onClick={[Function]}
            title="permission_templates.select_to_delete"
          >
            delete
          </Button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders optionrenderer and singlevaluerenderer: option renderer 1`] = `
<Option
  data={
    {
      "value": "val",
    }
  }
>
  <div
    className="display-flex-center"
  >
    <QualifierIcon
      className="little-spacer-right"
      qualifier="val"
    />
  </div>
</Option>
`;

exports[`renders optionrenderer and singlevaluerenderer: single value renderer 1`] = `
<SingleValue
  data={
    {
      "value": "val",
    }
  }
>
  <div
    className="display-flex-center"
  >
    <QualifierIcon
      className="little-spacer-right"
      qualifier="val"
    />
  </div>
</SingleValue>
`;
